utils/fork-tree/src/lib.rs

```diff
@@ -624,6 +624,35 @@ where
 			(None, false) => Ok(FinalizationResult::Unchanged),
 		}
 	}
+
+	/// Remove from the tree some nodes (and their subtrees) using a `predicate`.
+	/// The `predicate` should return a tuple where the first `bool` indicates
+	/// if a node and its subtree should be removed, the second `Option<bool>`
+	/// indicates if the filtering operation should be stopped.
+	/// If `Some(false)` stop is requested only for the current node subtree,
+	/// if `Some(true)` stop is requested for the full filtering operation.
+	/// Tree traversal is performed using an pre-order depth-first search.
+	/// An iterator over all the pruned nodes is returned.
+	pub fn filter<F>(&mut self, predicate: &mut F) -> impl Iterator<Item = (H, N, V)>
```

Comment:
Would you hate me if I ask you to combine this `(bool, Option<bool>)` into an `enum TreeFiltering` similar to how `poll` and `cmp` has their own result types?

Reply:
Actually I would love to use an enum. But we require the following enum variants:

```
enum TreeFiltering {
    RemoveAndContinue,
    RemoveAndStopThisBranch, // << This can be suppressend since Remove imples StopThisBranch
    RemoveAndStopAll,
    NoRemoveAndContinue,
    NoRemoveAndStopThisBranch,
    NoRemoveAndStopAll,
}
```
I didn't do that because I found it a bit ugly :-)

Reply:
Yeah, it seems the tuple pair is justified then. How about an

```rust
enum Traversal {
  Continue,
  Ignore,
  About,
}
```

and a return type of `(bool, Traversal)` to make the intent more clear?

Reply:
Seems reasonable

Reply:
This function looks very similar to https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter.
* naming suggestion: `drain_filter` is better than `filter`, because the latter usually conventionally means a non-mutating method
* `Vec::drain_filter` is implemented in O(n), while this is O(n^2). Might not be a problem in practice, but something to be aware of (and maybe document)